### PR TITLE
V4

### PR DESCRIPTION
**Pull request by LucianBuzzo**
Upgrades to react 16 and version bumps all dependencies.
Tooltips have been refactored to use an inline property.
Some components have been converted to TypeScript.

### CODE REVIEW

File: src/components/Badge.tsx [hunk 1]
```diff
@@ -37,5 +38,5 @@ export default withTheme(
 				{text}
 			</BaseBadge>
 		);
-	},
-);
+	}),
+) as React.ComponentClass<BadgeProps>;
```
Comment (pimterry) [inline src/components/Badge.tsx]: Why is this cast required?

File: src/components/PineTypes/Boolean.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import * as React from 'react'
 import Select from '../Select'
-import * as isBoolean from 'lodash/isBoolean'
-import * as isString from 'lodash/isString'
+const isBoolean = require('lodash/isBoolean')
+const isString = require('lodash/isString')
```
Comment (pimterry) [inline src/components/PineTypes/Boolean.js]: Why do we need this?

File: src/hoc.tsx [hunk 3]
```diff
@@ -5,6 +5,9 @@ import styled from 'styled-components';
 import { color, fontSize, space, width } from 'styled-system';
 import tag from 'tag-hoc';
 import blacklist from './blacklist';
+import { Tooltips } from './tooltips';
+
+const tt = new Tooltips();
```
Comment (pimterry) [inline src/hoc.tsx]: Nit: `tooltip` would be a little clearer

File: src/tooltips.ts [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+							CLICK_TIMEOUT,
+						);
+					};
+					props.onClick = (e: Event) => showFn(e) || hideFn() || oldFn(e);
+				} else {
+					const oldMEFn = props.onMouseEnter || noop;
+					props.onMouseEnter = (e: Event) => showFn(e) || oldMEFn(e);
+
+					const oldMLFn = props.onMouseLeave || noop;
+					props.onMouseLeave = (e: Event) => this.hide() || oldMLFn(e);
```
Comment (pimterry) [inline src/tooltips.ts]: Here (and twice more above) there's bit of clever `() => a || b` to run `a` & then `b`. That works, but it's a little confusing and a little risky - if we accidentally return something truthy in any `a`, then `b` will never run.

Could these just be `() => { a(); b(); }` instead?

File: src/stories/Tooltip.js [hunk 5]
```diff
@@ -1,37 +1,179 @@
...
+- Modal
+- Navbar
+- PineTypes
+- Provider
+- ProgressBar
+- Select
+- Table
+- Text
+- Textarea
+*/
```
Comment (pimterry) [inline src/stories/Tooltip.js]: Leftover todo - are we doing this? If not now, it'd be nice to move it into an issue instead.

File: src/tooltips.ts [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+	private tooltipElementArrow: HTMLDivElement;
+	private hideTimeout: number;
+	private initialised: boolean = false;
+
+	initialiseElements() {
+		if (this.initialised || !document || !document.body) {
+			return;
+		}
+
+		const tooltipElement = document.createElement('div');
```
Comment (pimterry) [inline src/tooltips.ts]: This is interesting. Why can't we use React + styled components to do all this tooltip stuff?

You probably don't want to put it in the downstream tree, but you could define a component here and separately call ReactDOM.render() with it to put it in the page yourself. I think that should still let you do this The React Way, and make a lot of this much simpler.

Comment (LucianBuzzo) [reply src/tooltips.ts]: Good idea

File: src/components/Badge.tsx [hunk 7]
```diff
@@ -22,7 +23,7 @@ const BaseBadge = styled(Box)`
 `;
 
 export default withTheme(
-	({ small, text, theme, ...props }: ThemedBadgeProps) => {
+	hoc(({ small, text, theme, ...props }: ThemedBadgeProps) => {
```
[the remark this answers is not in the document]
Comment (LucianBuzzo) [reply src/components/Badge.tsx]: +1 for `asRendition`

File: src/components/Badge.tsx (hunk 1 above)
Comment (LucianBuzzo) [reply src/components/Badge.tsx]: To keep me sane! I think we just started using this pattern to get consistent output between `compose` and `withTheme`. The styled components typings are not great..

File: src/components/PineTypes/Boolean.js (hunk 2 above)
Comment (LucianBuzzo) [reply src/components/PineTypes/Boolean.js]: I think jest was choking on the import syntax here

**Pull request merged**